src/tokenizer/shingle_filter.rs

```diff
@@ -0,0 +1,207 @@
+        FixedShingleFilter::new_with_separator(size, '_')
+    }
+
+    pub fn new_with_separator(size: usize, separator: char) -> Self {
+        FixedShingleFilter { size, separator }
+    }
+}
+
+impl TokenFilter for FixedShingleFilter {
+    fn transform<'a>(&self, token_stream: BoxTokenStream<'a>) -> BoxTokenStream<'a> {
+        BoxTokenStream::from(ShingleFilterStream {
+            size: self.size,
+            separator: self.separator,
+            buffer: String::with_capacity(100),
+            pendings: Vec::with_capacity(self.size - 1),
+            tail: token_stream,
+            original_token_position: 0,
+        })
+    }
+}
+
+pub struct ShingleFilterStream<'a> {
+    size: usize,
+    separator: char,
+    buffer: String,
+    pendings: Vec<Token>,
+    tail: BoxTokenStream<'a>,
+    original_token_position: usize,
+}
+
+impl<'a> ShingleFilterStream<'a> {
+    fn is_window_filled(&self) -> bool {
+        self.pendings.len() == self.size - 1
+    }
+
+    fn push_token(&mut self, mut token: Token) {
+        if self.is_window_filled() {
+            self.pendings.remove(0);
+        }
+        token.position = self.original_token_position;
+        self.pendings.push(token);
+    }
+}
+
+fn make_shingle(token: &mut Token, pendings: &Vec<Token>, buffer: &mut String, separator: char) {
+    buffer.clear();
+    buffer.push_str(token.text.as_str());
+
+    // set text
+    token.text.clear();
+    for pend_token in pendings {
+        token.text.push_str(pend_token.text.as_str());
+        token.text.push(separator);
+    }
+    token.text.push_str(buffer.as_str());
+
+    // set offset and position
+    if let Some(first_token) = pendings.get(0) {
+        token.offset_from = first_token.offset_from;
+        token.position = first_token.position;
+    }
+}
+
+impl<'a> TokenStream for ShingleFilterStream<'a> {
+    fn advance(&mut self) -> bool {
+        while !self.is_window_filled() {
+            if self.tail.advance() {
+                self.push_token(self.tail.token().clone());
+                self.original_token_position += 1;  // This is not compatible with StopWordFilter
+            } else {
+                return false;
+            }
+        }
+        if self.tail.advance() {
+            let next_token = self.tail.token().clone();
+            make_shingle(
+                &mut self.tail.token_mut(),
+                &self.pendings,
+                &mut self.buffer,
+                self.separator,
+            );
+            self.push_token(next_token);
+            self.original_token_position += 1;
+            return true;
+        }
+        false
+    }
+
+    fn token(&self) -> &Token {
+        self.tail.token()
+    }
+
+    fn token_mut(&mut self) -> &mut Token {
+        self.tail.token_mut()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::tokenizer::{FixedShingleFilter, SimpleTokenizer, TextAnalyzer, Token};
+
+    #[test]
+    fn test_shingle_bigram() {
```

Comment:
The code handles this case just fine but can you add a test where we have less than "window size" tokens to begin with?
(The expected result is that we should emit not token).

Reply:
And one test with the empty string too.